### PR TITLE
[1.13.x] Fix race accessing plugin storage map

### PR DESCRIPTION
**Pull request by cpuguy83**
`plugins.GetAll()` was not locking the plugin map when checking if a
plugin exists, this can cause a race and potentially a panic if another
goroutine is trying to load a plugin into the map at the same time.

Also fixes a race during activation where a plugin inserts itself into
the plugin map but does not check if something else is already there.
This is already checked before trying to activate the plugin, however
the map lock is not held for this entire period, so other plugins may be
loaded during this time.
To fix, before inserting the plugin into the map, check if one with the
same name already exists and use that instead.

Signed-off-by: Brian Goff <cpuguy83@gmail.com>
(cherry picked from commit 9080c92a57c2549ca1695390fc0f634b73a21d7a)
Signed-off-by: Brian Goff <cpuguy83@gmail.com>

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by cpuguy83** [comment]
Counterpart to #29893
ping @vieux @tiborvass

**Comment by tiborvass** [comment]
LGTM

**Comment by vieux** [comment]
LGTM

**Comment by vieux** [comment]
@cpuguy83 test failures, I think they are related

**Comment by cpuguy83** [comment]
This is all fixed.

File: pkg/plugins/plugins.go [hunk 1]
```diff
@@ -221,6 +221,10 @@ func loadWithRetry(name string, retry bool) (*Plugin, error) {
 		}
 
 		storage.Lock()
+		if pl, exists := storage.plugins[name]; exists {
+			storage.Unlock()
+			return pl, pl.activate()
```
Comment (thaJeztah) [inline pkg/plugins/plugins.go]: If `activate()` fails, should the plugin be removed (like the code below does?

Comment (thaJeztah) [inline pkg/plugins/plugins.go]: discussed on Slack with @cpuguy83, and not worth changing here

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
all green, merging

**Pull request merged**